server/storage/database.go

```diff
@@ -1,177 +1,165 @@
+// See server/storage/models.go
+type SQLStorage struct {
+	gorm.DB
 }
 
-// NewMySQLStorage is a convenience method to create a MySQLStorage
-func NewMySQLStorage(db *sql.DB) *MySQLStorage {
-	return &MySQLStorage{
-		DB: *db,
+// NewSQLStorage is a convenience method to create a SQLStorage
+func NewSQLStorage(dialect string, args ...interface{}) (*SQLStorage, error) {
+	gormDB, err := gorm.Open(dialect, args...)
+	if err != nil {
+		return nil, err
 	}
+	return &SQLStorage{
+		DB: gormDB,
+	}, nil
 }
 
-// UpdateCurrent updates multiple TUF records in a single transaction.
-// Always insert a new row. The unique constraint will ensure there is only ever
-func (db *MySQLStorage) UpdateCurrent(gun string, update MetaUpdate) error {
-	checkStmt := "SELECT count(*) FROM `tuf_files` WHERE `gun`=? AND `role`=? AND `version`>=?;"
-	insertStmt := "INSERT INTO `tuf_files` (`gun`, `role`, `version`, `data`) VALUES (?,?,?,?);"
+// UpdateCurrent updates a single TUF.
+func (db *SQLStorage) UpdateCurrent(gun string, update MetaUpdate) error {
+	// ensure we're not inserting an immediately old version - can't use the
+	// struct, because that only works with non-zero values, and Version
+	// can be 0.
+	exists := db.Where("gun = ? and role = ? and version >= ?",
+		gun, update.Role, update.Version).First(&TUFFile{})
 
-	// ensure we're not inserting an immediately old version
-	row := db.QueryRow(checkStmt, gun, update.Role, update.Version)
-	var exists int
-	err := row.Scan(&exists)
-	if err != nil {
-		return err
-	}
-	if exists != 0 {
+	if !exists.RecordNotFound() {
 		return &ErrOldVersion{}
 	}
 
-	// attempt to insert. Due to race conditions with the check this could fail.
-	// That's OK, we're doing first write wins. The client will be messaged it
-	// needs to rebase.
-	_, err = db.Exec(insertStmt, gun, update.Role, update.Version, update.Data)
-	if err != nil {
-		if err, ok := err.(*mysql.MySQLError); ok {
-			if err.Number == 1022 { // duplicate key error
-				return &ErrOldVersion{}
-			}
-		}
-		// need to check error type for duplicate key exception
-		// and return ErrOldVersion if duplicate
-		return err
-	}
-	return nil
+	return db.Create(&TUFFile{
+		Gun:     gun,
+		Role:    update.Role,
+		Version: update.Version,
+		Data:    update.Data,
+	}).Error
 }
 
 // UpdateMany atomically updates many TUF records in a single transaction
-func (db *MySQLStorage) UpdateMany(gun string, updates []MetaUpdate) error {
-	checkStmt := "SELECT count(*) FROM `tuf_files` WHERE `gun`=? AND `role`=? AND `version`>=?;"
-	insertStmt := "INSERT INTO `tuf_files` (`gun`, `role`, `version`, `data`) VALUES (?,?,?,?);"
+func (db *SQLStorage) UpdateMany(gun string, updates []MetaUpdate) error {
+	tx := db.Begin()
+	if tx.Error != nil {
+		return tx.Error
+	}
 
-	tx, err := db.Begin()
-	if err != nil {
+	rollback := func(err error) error {
+		query := tx.Rollback()
+		if query.Error != nil {
+			logrus.Panic("Failed on Tx rollback with error: ", err.Error())
```

Comment:
Also shouldn't panic once we're into real runtime code (as opposed to initialization code). It'll take down the server.